### ISSUE TITLE
Better OpenSSL v1.02q Compatibility

### ISSUE THREAD

**Issue by hlsantos** [openssl/openssl#8575]
Is there is a consolidated, accumulated guide, documentation, code, etc, that will ease the migration, porting pains from v1.0.2q to v1.1.1x?  

The code change requirements are extensive.  It is getting to a point where removal of extensive code without any workaround or explanation is overbearing and is putting a lot of pressure on me to put this v1.1.1 migration project on the back burner.  I think the EOL for 1.0.2g is pretty ambitious when backward code compatibility has been taken rather lightly than I would expect.

I'm doing my best to do this compatibility porting work to get our long time Openssl implementation working again at the functional and rudimentary SSL level, but the documentation for what has been removed, the whys, whether it is safe to remove, what can replaced it, if anything, etc is very lacking.   In some cases, removal was done at the API coding level, but it may be still available via the apps using internal headers, and the functions may still be documented on line.  So what gives?  Do you need these parts or not?  Yet,  these parts are hidden or removed from the APIs creating a greater "spaghetti code" situation using "internal headers."  

I am trying to remain positive about this and would be a great pleasure to assist where I can, but I don't like stepping on people's toes.   Yet, the Openssl Authors are stepping on "implementation toes" in a rather major and drastic way.

There has be a better way to do this Openssl v1.1.1 work without breaking so many parts and systems, and worst, its almost impossible to recompile existing production without a major expensive rewrite.

Thanks for listening.

**Comment by richsalz** [comment]
The work to make structures opaque, which is causing the problems, happened in the 1.1.0 release which was in August of 2016.  See https://www.openssl.org/news/openssl-1.1.0-notes.html

**Comment by hlsantos** [comment]
Yes , I was aware there would be changes and I waited for stability.  Needed to be planned.  Three years later, I would expect the technical porting, migration and code change requirements to be reasonably documented, in particular in wrapping new reader/writer methods to common struct fields used for common and basic SSL usages. 

There can perhaps be a central table and cross reference for this.   I just spent a few hours trying to find an access function for sk_X509_ATTRIBUTE_num (req->req_info->attributes) realizing it probably isn't needed anymore and now d2i_RSA_NET() using for loading and checking private keys.  

Trying to backtrack this by looking at the apps *.c code comparing the old with the new, and much of this has changed or removed,  i.e. the load_netscrape_key() function was removed from apps\apps.c making it very hard to follow what needs to be done,

**Comment by mspncp** [comment]
At the time when the structures were made opaque, there was a lot of discussion in the community about a joint effort to create an officially supported compatibility layer (either by the OpenSSL project or the community of OpenSSL users), which would facitilitate porting from 1.0.2 to 1.1.0. 

Unfortunately, no consensus was never reached about how this should be achieved. And in the end, most of the downstream projects had rewritten their code and/or introduced their own compatibility layers anyway, so there is no such official compatibility layer, unfortunately. At least none I know of.

If you are interested in the details, you can read the following thread, which started in Jan 2016 and was resumed in Nov 2016:

[[openssl-dev] Backporting opaque struct getter/setter functions  (Jan 2016)](https://mta.openssl.org/pipermail/openssl-dev/2016-January/004005.html)

[[openssl-dev] Backporting opaque struct getter/setter functions  (Nov 2016)](https://mta.openssl.org/pipermail/openssl-dev/2016-November/008677.html)

**Comment by hlsantos** [comment]
Thanks for your input. I will take the time to read the threads.  

My final comments on this, then closing it.

I think what I see is that legacy functionality were removed without significant accounting for established products who may have  legacy conversion features , i.e. IIS4/IIS5 conversion options that our SSL setup/configuration tool offers.   Its probably not needed any more, but it could have been put into a legacy API without removing it with "deprecation" compiler macros..    

The problem is more about removal without replacement/workaround/documentation.   Not so much the replacing struct field access with "assessors" functions as long as the documentation is provided, this is actually pretty simple process.   If a mapping table was provided, it can be used as a part of a conversion "lint" tool.

Thanks

**Comment by kroeckx** [comment]
For the opaque things, most of that should be covered by:
https://wiki.openssl.org/index.php/OpenSSL_1.1.0_Changes

I think that if we removed old formats, it's because they are
insecure and we think they haven't been used in the last 10 years.
